src/Service/S3/src/S3Client.php

```diff
@@ -652,6 +652,27 @@ public function uploadPart($input): UploadPartOutput
         return new UploadPartOutput($response);
     }
 
+    protected function getEndpoint(string $uri, array $query, ?string $region): string
+    {
+        $uriParts = \explode('/', $uri, 3);
+        $bucket = $uriParts[1] ?? '';
+        $bucketLen = \strlen($bucket);
+        $configuration = $this->getConfiguration();
+
+        /** @psalm-suppress PossiblyNullArgument */
+        if (
+        $bucketLen < 3 || $bucketLen > 63
+        || filter_var($bucket, \FILTER_VALIDATE_IP) // Cannot look like an IP address
+        || !preg_match('/^[a-z0-9]([a-z0-9\-]*[a-z0-9])?$/', $bucket) // Bucket cannot have dot (because of TLS)
+        || filter_var(\parse_url($configuration->get('endpoint'), \PHP_URL_HOST), \FILTER_VALIDATE_IP) // Custom endpoint cannot look like an IP address @phpstan-ignore-line
+        || \filter_var($configuration->get('s3PathStyleEndpoint'), \FILTER_VALIDATE_BOOLEAN)
+        ) {
+            return parent::getEndpoint($uri, $query, $region);
+        }
+
+        return \preg_replace(';https?://;', '$0' . $uriParts[1] . '.', parent::getEndpoint('/' . ($uriParts[2] ?? ''), $query, $region));
```

Comment:
using `;` as a pattern delimiter looks really odd to me. I've never seen that. How about `|` which seems to be a lot more common>

Reply:
thanks for the suggestion